File: orcastork/adapters/memory/store.py
"""In-memory ``DataPointStore`` — deterministic, infra-free, fenced keyed-merge store.

Per session it keeps each identity's DataPoint plus the revisions at which it was first
added and last freshened, so ``change_set_since`` can cleanly split added vs updated. The
session revision advances once per write that actually changed something. Writes are
guarded by the highest epoch seen for the session (lower epoch → ``StaleEpochError``).
"""

from __future__ import annotations

from collections.abc import Iterable, Sequence
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any

from ...datapoints import BaseDataPoint, DataPointView, identity_key
from ...exceptions import StaleEpochError
from ...ids import Epoch, OperatorId, Revision, SessionId
from ...ports.change_set import ChangeSet
from ...ports.datapoint_store import EFFECT_COMMITTED, EffectClaim, effect_pending_state


@dataclass
class _Entry:
    data_point: BaseDataPoint[Any]
    added_rev: int
    updated_rev: int


@dataclass
class _SessionState:
    entries: dict[tuple[str, Any], _Entry] = field(default_factory=dict)
    revision: int = 0
    max_epoch: int = 0
    watermarks: dict[OperatorId, int] = field(default_factory=dict)
    effects: dict[str, str] = field(default_factory=dict)  # effect_key → 'pending:<epoch>' | 'committed'
    deadline: datetime | None = None  # the persisted wall-clock session deadline (set once, rehydrated on resume)
    flow_fingerprint: str | None = None  # the latest flow that drove the session (drift detection on resume)


class InMemoryDataPointStore:
    def __init__(self) -> None:
        self._sessions: dict[SessionId, _SessionState] = {}

    def _state(self, session_id: SessionId) -> _SessionState:
        return self._sessions.setdefault(session_id, _SessionState())

    @staticmethod
    def _guard_epoch(state: _SessionState, epoch: Epoch) -> None:
        # Ownership is the lock's job; recording the highest ACCEPTED epoch on EVERY guarded
        # write is what completes stale-writer rejection — a session whose first mutation is an
        # effect claim or a session-meta write must still fence a later lower-epoch writer.
        if epoch < state.max_epoch:
            raise StaleEpochError(f'epoch {epoch} is stale (current {state.max_epoch})')
        state.max_epoch = epoch

    async def write(
        self, session_id: SessionId, data_points: Iterable[BaseDataPoint[Any]], *, epoch: Epoch
    ) -> Revision:
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        next_revision = state.revision + 1
        changed = False
        for data_point in data_points:
            key = identity_key(data_point)
            existing = state.entries.get(key)
            if existing is None:
                state.entries[key] = _Entry(data_point, added_rev=next_revision, updated_rev=next_revision)
                changed = True
            elif data_point.last_retrieved > existing.data_point.last_retrieved:
                existing.data_point = existing.data_point.reobserved(data_point.last_retrieved)
                existing.updated_rev = next_revision
                changed = True
        if changed:
            state.revision = next_revision
        return Revision(state.revision)

    async def apply_resolved(
        self,
        session_id: SessionId,
        *,
        added: Sequence[BaseDataPoint[Any]],
        updated: Sequence[BaseDataPoint[Any]],
        epoch: Epoch,
    ) -> Revision:
        # The caller (a sole mutator) already keyed-merged, so this applies blindly: new
        # identities land with both stamps at the batch revision, merged identities overwrite
        # their entry and advance only the updated stamp — exactly the stamping `write` does.
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        if not added and not updated:
            return Revision(state.revision)
        next_revision = state.revision + 1
        for data_point in added:
            state.entries[identity_key(data_point)] = _Entry(
                data_point, added_rev=next_revision, updated_rev=next_revision
            )
        for data_point in updated:
            key = identity_key(data_point)
            existing = state.entries.get(key)
            # An update of an unknown identity still lands (blind application, mirroring the
            # Redis HSET) — it simply carries no added stamp, like a field set by mode 'u'.
            added_rev = 0 if existing is None else existing.added_rev
            state.entries[key] = _Entry(data_point, added_rev=added_rev, updated_rev=next_revision)
        state.revision = next_revision
        return Revision(next_revision)

    async def snapshot(self, session_id: SessionId) -> DataPointView:
        return DataPointView(entry.data_point for entry in self._state(session_id).entries.values())

    async def revision(self, session_id: SessionId) -> Revision:
        return Revision(self._state(session_id).revision)

    async def change_set_since(self, session_id: SessionId, since: Revision) -> ChangeSet:
        state = self._state(session_id)
        added = tuple(e.data_point for e in state.entries.values() if e.added_rev > since)
        updated = tuple(e.data_point for e in state.entries.values() if e.added_rev <= since < e.updated_rev)
        return ChangeSet(added=added, updated=updated)

    async def get_watermark(self, session_id: SessionId, operator_id: OperatorId) -> Revision | None:
        watermark = self._state(session_id).watermarks.get(operator_id)
        return None if watermark is None else Revision(watermark)

    async def set_watermark(
        self, session_id: SessionId, operator_id: OperatorId, revision: Revision, *, epoch: Epoch
    ) -> None:
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        state.watermarks[operator_id] = revision

    async def claim_effect(
        self, session_id: SessionId, effect_key: str, *, epoch: Epoch, reclaim_stale: bool
    ) -> EffectClaim:
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        stored = state.effects.get(effect_key)
        if stored is None:
            state.effects[effect_key] = effect_pending_state(epoch)
            return EffectClaim.ACQUIRED
        if stored == EFFECT_COMMITTED:
            return EffectClaim.ALREADY_COMMITTED
        if stored == effect_pending_state(epoch):
            return EffectClaim.PENDING_SAME_EPOCH
        if reclaim_stale:
            state.effects[effect_key] = effect_pending_state(epoch)
            return EffectClaim.ACQUIRED
        return EffectClaim.PENDING_STALE_EPOCH

    async def commit_effect(self, session_id: SessionId, effect_key: str, *, epoch: Epoch) -> None:
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        # Only this epoch's own pending mark transitions; 'committed' stays as-is (idempotent), and
        # any other state is left untouched — never fabricate 'committed' for an unowned claim.
        if state.effects.get(effect_key) in (effect_pending_state(epoch), EFFECT_COMMITTED):
            state.effects[effect_key] = EFFECT_COMMITTED

    async def revert_effect(self, session_id: SessionId, effect_key: str, *, epoch: Epoch) -> None:
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        # Deletes ONLY this epoch's own pending mark: never 'committed' (the effect DID run) and
        # never another epoch's pending (resolved by its owner, or by a successor's recovery policy).
        if state.effects.get(effect_key) == effect_pending_state(epoch):
            del state.effects[effect_key]

    async def get_effect_state(self, session_id: SessionId, effect_key: str) -> str | None:
        return self._state(session_id).effects.get(effect_key)

    async def get_session_deadline(self, session_id: SessionId) -> datetime | None:
        return self._state(session_id).deadline

    async def set_session_deadline(self, session_id: SessionId, deadline: datetime, *, epoch: Epoch) -> None:
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        state.deadline = deadline

    async def get_flow_fingerprint(self, session_id: SessionId) -> str | None:
        return self._state(session_id).flow_fingerprint

    async def set_flow_fingerprint(self, session_id: SessionId, fingerprint: str, *, epoch: Epoch) -> None:
        state = self._state(session_id)
        self._guard_epoch(state, epoch)
        state.flow_fingerprint = fingerprint

File: typescript/src/orcastork/adapters/memory/store.ts
/**
 * In-memory `DataPointStore` — deterministic, infra-free, fenced keyed-merge store.
 *
 * Per session it keeps each identity's DataPoint plus the revisions at which it was first added
 * and last freshened, so `changeSetSince` can cleanly split added vs updated. The session revision
 * advances once per write that actually changed something. Writes are guarded by the highest epoch
 * seen for the session (lower epoch → `StaleEpochError`).
 *
 * @module
 */

import type { AnyDataPoint } from '../../datapoints/index.js';
import { DataPointView, identityKey } from '../../datapoints/index.js';
import { StaleEpochError } from '../../exceptions.js';
import type { Epoch, OperatorId, Revision, SessionId } from '../../ids.js';
import { Revision as toRevision } from '../../ids.js';
import { ChangeSet } from '../../ports/change_set.js';
import type { ApplyResolvedOptions, ClaimEffectOptions, DataPointStore } from '../../ports/datapoint_store.js';
import { EFFECT_COMMITTED, EffectClaim, effectPendingState } from '../../ports/datapoint_store.js';

/** One stored identity: the DataPoint plus the revisions that stamp it added / last freshened. */
interface Entry {
  dataPoint: AnyDataPoint;
  addedRev: number;
  updatedRev: number;
}

/** Everything one session owns: its blackboard and its epoch-fenced meta state. */
interface SessionState {
  readonly entries: Map<string, Entry>;
  revision: number;
  maxEpoch: number;
  readonly watermarks: Map<OperatorId, number>;

  /** effect key → `pending:<epoch>` | `committed`. */
  readonly effects: Map<string, string>;

  /** The persisted wall-clock session deadline (set once, rehydrated on resume). */
  deadline: Date | null;

  /** The latest flow that drove the session (drift detection on resume). */
  flowFingerprint: string | null;
}

const newSessionState = (): SessionState => ({
  entries: new Map<string, Entry>(),
  revision: 0,
  maxEpoch: 0,
  watermarks: new Map<OperatorId, number>(),
  effects: new Map<string, string>(),
  deadline: null,
  flowFingerprint: null,
});

/**
 * Reject a superseded writer and record the highest ACCEPTED epoch.
 *
 * Ownership is the lock's job; recording the highest accepted epoch on EVERY guarded write is what
 * completes stale-writer rejection — a session whose first mutation is an effect claim or a
 * session-meta write must still fence a later lower-epoch writer.
 */
const guardEpoch = (state: SessionState, epoch: Epoch): void => {
  if (epoch < state.maxEpoch) {
    throw new StaleEpochError(`epoch ${epoch} is stale (current ${state.maxEpoch})`);
  }
  state.maxEpoch = epoch;
};

/** The blackboard every other in-memory adapter is written against. */
export class InMemoryDataPointStore implements DataPointStore {
  private readonly sessions = new Map<SessionId, SessionState>();

  private stateOf(sessionId: SessionId): SessionState {
    const known = this.sessions.get(sessionId);
    if (known !== undefined) {
      return known;
    }
    const created = newSessionState();
    this.sessions.set(sessionId, created);
    return created;
  }

  public async write(
    sessionId: SessionId,
    dataPoints: Iterable<AnyDataPoint>,
    options: { readonly epoch: Epoch },
  ): Promise<Revision> {
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    const nextRevision = state.revision + 1;
    let changed = false;
    for (const dataPoint of dataPoints) {
      const key = identityKey(dataPoint);
      const existing = state.entries.get(key);
      if (existing === undefined) {
        state.entries.set(key, { dataPoint, addedRev: nextRevision, updatedRev: nextRevision });
        changed = true;
      } else if (dataPoint.lastRetrieved.getTime() > existing.dataPoint.lastRetrieved.getTime()) {
        existing.dataPoint = existing.dataPoint.reobserved(dataPoint.lastRetrieved);
        existing.updatedRev = nextRevision;
        changed = true;
      }
    }
    if (changed) {
      state.revision = nextRevision;
    }
    return toRevision(state.revision);
  }

  public async applyResolved(sessionId: SessionId, options: ApplyResolvedOptions): Promise<Revision> {
    // The caller (a sole mutator) already keyed-merged, so this applies blindly: new identities
    // land with both stamps at the batch revision, merged identities overwrite their entry and
    // advance only the updated stamp — exactly the stamping `write` does.
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    if (options.added.length === 0 && options.updated.length === 0) {
      return toRevision(state.revision);
    }
    const nextRevision = state.revision + 1;
    for (const dataPoint of options.added) {
      state.entries.set(identityKey(dataPoint), {
        dataPoint,
        addedRev: nextRevision,
        updatedRev: nextRevision,
      });
    }
    for (const dataPoint of options.updated) {
      const key = identityKey(dataPoint);
      const existing = state.entries.get(key);
      // An update of an unknown identity still lands (blind application, mirroring the Redis HSET)
      // — it simply carries no added stamp, like a field set by mode 'u'.
      const addedRev = existing === undefined ? 0 : existing.addedRev;
      state.entries.set(key, { dataPoint, addedRev, updatedRev: nextRevision });
    }
    state.revision = nextRevision;
    return toRevision(nextRevision);
  }

  public async snapshot(sessionId: SessionId): Promise<DataPointView> {
    return new DataPointView([...this.stateOf(sessionId).entries.values()].map((entry) => entry.dataPoint));
  }

  public async revision(sessionId: SessionId): Promise<Revision> {
    return toRevision(this.stateOf(sessionId).revision);
  }

  public async changeSetSince(sessionId: SessionId, since: Revision): Promise<ChangeSet> {
    const entries = [...this.stateOf(sessionId).entries.values()];
    return new ChangeSet({
      added: entries.filter((entry) => entry.addedRev > since).map((entry) => entry.dataPoint),
      updated: entries
        .filter((entry) => entry.addedRev <= since && since < entry.updatedRev)
        .map((entry) => entry.dataPoint),
    });
  }

  public async getWatermark(sessionId: SessionId, operatorId: OperatorId): Promise<Revision | null> {
    const watermark = this.stateOf(sessionId).watermarks.get(operatorId);
    return watermark === undefined ? null : toRevision(watermark);
  }

  public async setWatermark(
    sessionId: SessionId,
    operatorId: OperatorId,
    revision: Revision,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    state.watermarks.set(operatorId, revision);
  }

  public async claimEffect(
    sessionId: SessionId,
    effectKey: string,
    options: ClaimEffectOptions,
  ): Promise<EffectClaim> {
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    const stored = state.effects.get(effectKey);
    if (stored === undefined) {
      state.effects.set(effectKey, effectPendingState(options.epoch));
      return EffectClaim.ACQUIRED;
    }
    if (stored === EFFECT_COMMITTED) {
      return EffectClaim.ALREADY_COMMITTED;
    }
    if (stored === effectPendingState(options.epoch)) {
      return EffectClaim.PENDING_SAME_EPOCH;
    }
    if (options.reclaimStale) {
      state.effects.set(effectKey, effectPendingState(options.epoch));
      return EffectClaim.ACQUIRED;
    }
    return EffectClaim.PENDING_STALE_EPOCH;
  }

  public async commitEffect(
    sessionId: SessionId,
    effectKey: string,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    // Only this epoch's own pending mark transitions; 'committed' stays as-is (idempotent), and any
    // other state is left untouched — never fabricate 'committed' for an unowned claim.
    const stored = state.effects.get(effectKey);
    if (stored === effectPendingState(options.epoch) || stored === EFFECT_COMMITTED) {
      state.effects.set(effectKey, EFFECT_COMMITTED);
    }
  }

  public async revertEffect(
    sessionId: SessionId,
    effectKey: string,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    // Deletes ONLY this epoch's own pending mark: never 'committed' (the effect DID run) and never
    // another epoch's pending (resolved by its owner, or by a successor's recovery policy).
    if (state.effects.get(effectKey) === effectPendingState(options.epoch)) {
      state.effects.delete(effectKey);
    }
  }

  public async getEffectState(sessionId: SessionId, effectKey: string): Promise<string | null> {
    return this.stateOf(sessionId).effects.get(effectKey) ?? null;
  }

  public async getSessionDeadline(sessionId: SessionId): Promise<Date | null> {
    return this.stateOf(sessionId).deadline;
  }

  public async setSessionDeadline(
    sessionId: SessionId,
    deadline: Date,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    state.deadline = deadline;
  }

  public async getFlowFingerprint(sessionId: SessionId): Promise<string | null> {
    return this.stateOf(sessionId).flowFingerprint;
  }

  public async setFlowFingerprint(
    sessionId: SessionId,
    fingerprint: string,
    options: { readonly epoch: Epoch },
  ): Promise<void> {
    const state = this.stateOf(sessionId);
    guardEpoch(state, options.epoch);
    state.flowFingerprint = fingerprint;
  }
}
